Add Login preferences block.

abdullah-kasim opened the pull request:
<!--
Link a related Github/Linear issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword. Consider using "part of" instead.
-->

Part of DOTDASH-269

## Proposed Changes

1. Adds the Preference page.
2. Adds the Preference > Login Preference block.
3. Port previous integration to Hosting Dashboard style of APIs. This refers to the `/me/settings` endpoint as well as the `/me/preferences` endpoint. The latter already exists, hence I only needed to add the additional types, but the former does not exist.

### Screens

#### Full page

<img width="3368" height="2352" alt="CleanShot 2025-09-16 at 16 34 40@2x" src="https://github.com/user-attachments/assets/62108467-977c-4696-a31d-d5087aecdef5" />

#### Detail of the open select ComboBox
<img width="1380" height="1274" alt="CleanShot 2025-09-16 at 16 35 05@2x" src="https://github.com/user-attachments/assets/d2f40791-064c-4f7d-89d6-0815098107ea" />


## Why are these changes being made?
<!--
It's easy to see what a PR does but much harder to find out why it was made,
particularly when researching old changes in history. Record an explanation of
the motivation behind this change and how it will help.
-->

These changes are made for:

1. To add the Preference page as part of the dashboard revamp.
2. To add the Login preference block.

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

1. Start the dev environment `yarn start`
2. Go to http://calypso.localhost:3000/v2/me/preferences
5. Click a field, a drop down, or similar.
6. Press Save.
7. Verify that we got a snackbar after saving is complete.

|  Before  | After |
|---------|------|
| N/A - page doesn't exist | <img width="1907" height="763" alt="image" src="https://github.com/user-attachments/assets/ee97671e-dc20-4ad1-80da-dbae10cf07b2" /> |

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [x] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [x] Have you tested accessibility for your changes? Ensure the feature remains usable with various user agents (e.g., browsers), interfaces (e.g., keyboard navigation), and assistive technologies (e.g., screen readers) (PCYsg-S3g-p2).
- [x] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

abdullah-kasim committed:
Add placeholder page for Preferences so we can start work on it.

abdullah-kasim committed:
Add skeleton for login preference's block

abdullah-kasim committed:
Add login preferences data

abdullah-kasim committed:
Fix skeleton to better match the design - still untested

abdullah-kasim committed:
Allrightey, add site selector and fix styling.

abdullah-kasim committed:
Move out save button logic into a hook

abdullah-kasim committed:
Refactor profile/index.tsx to use the newly refactored hook

abdullah-kasim committed:
Use the same save button logic as per the profile page

abdullah-kasim committed:
Match page layout with other pages (make it small, basically)

abdullah-kasim committed:
Merge remote-tracking branch 'origin/trunk' into add/preferences-login

abdullah-kasim commented on client/dashboard/me/preferences-login/index.tsx:
```diff
@@ -0,0 +1,166 @@
+
+					{ sites.length > 0 && (
+						<VStack>
+							<CustomSelectControl
```
CustomSelectControlV2 seems to be WIP and not exported by the root component, so I'll fallback to CustomSelectControl. That means no site icon, however.

abdullah-kasim commented on client/dashboard/me/profile/index.tsx:
```diff
@@ -136,13 +137,12 @@ export default function Profile() {
-	} else if ( mutation.isSuccess && ! isDirty ) {
-		saveButtonLabel = __( 'Saved!' );
-	}
+	const saveButtonState = useSaveButtonState( {
+		isSaving,
+		isSuccess: mutation.isSuccess,
+		isDirty,
+	} );
```
Refactored into a custom hook so that I can reuse it in other components. A lot of blocks will use this logic.

Alternatively we could create a `SaveButton` component though. What does this codebase prefer?

fushar replied:
I think for now we don't need to update the button text during form submission as we don't have consensus yet. See the most recent convo: p1747304091143159-slack-CNGQYA3B9

I know this is just a refactor for the existing code, but that was from the early prototype. 😄

fushar replied:
To give you some more context, this profile file was from an early prototype. Since then, we have refined how we do form submission. It would be good if you can simplify it to follow the most recent pattern, e.g.  https://github.com/Automattic/wp-calypso/blob/f26474a8e6311e6022b857234875ca7016f0676b/client/dashboard/sites/settings-php/index.tsx

p-jackson replied:
Just dropping in to say that following the most recent patterns means we shouldn't have a `Saved!` button label, we should use a snackbar to announce when saving is complete instead.

abdullah-kasim reviewed (commented):
I'm not sure how testing will go or when should I put this up for review (should it be after I write tests, or before?), but I'll put this up for review to get the feel for it.

abdullah-kasim committed:
Remove extra div and description - oops, that's not part of the design

abdullah-kasim committed:
Lint fixes

abdullah-kasim committed:
More lint fixes

taipeicoder commented on client/dashboard/me/preferences-login/index.tsx:
```diff
@@ -0,0 +1,165 @@
+								} }
+							/>
+							<Text variant="muted" style={ { fontSize: '13px', marginTop: '4px' } }>
+								{ __( "Choose the default site dashboard you'll see at login." ) }
```
As per pgz0xU-1S-p2 

```suggestion
								{ __( 'Choose the default site dashboard you’ll see at login.' ) }
```

fushar replied:
I think we should create a lint rule for this 🥲

taipeicoder commented on client/dashboard/me/preferences-login/index.tsx:
```diff
@@ -0,0 +1,165 @@
+							} }
+						/>
+						<Text variant="muted" style={ { fontSize: '13px', marginTop: '4px' } }>
+							{ __( "Select what you'll see by default when visiting WordPress.com." ) }
```
```suggestion
							{ __( 'Select what you’ll see by default when visiting WordPress.com.' ) }
```

abdullah-kasim commented:
Moving this back to draft as I implement the changes. Also I was just reading on DataForm 😅 - so switching to that.

abdullah-kasim committed:
Switch over to DataForm

abdullah-kasim committed:
useSaveButtonState is the old way of doing it - reverting the refactor.
We use snackbars now.

abdullah-kasim committed:
Add tests for preferences page

abdullah-kasim committed:
Add tests for the login preferences component

abdullah-kasim committed:
Merge remote-tracking branch 'origin/trunk' into add/preferences-login

abdullah-kasim committed:
Fix import path changes from merge

abdullah-kasim committed:
Use __next40pxDefeaultSize as per deprecation notice

abdullah-kasim committed:
Merge remote-tracking branch 'origin/trunk' into add/preferences-login

abdullah-kasim committed:
Change import paths due to changes in trunk

abdullah-kasim commented:
Ready for another round of review. I've added tests.

I contemplated adding `e2e's` as per `docs/testing.md` but that folder doesn't look updated, and recent pattern seems to add tests **near** the file that we're editing/adding, inside a `tests` folder with `.test.tsx` in the filename.

Happy to also update `testing.md` in a separate PR - just let me know!

p-jackson commented on packages/api-core/src/me-preferences/types.ts:
```diff
@@ -9,7 +9,15 @@ export type SitesViewPreferences = Partial< Omit< SitesView, 'type' | 'layout' >
+
 export interface UserPreferences {
 	'sites-view'?: SitesViewPreferences;
+	'login-preferences'?: LoginPreferences;
```
This doesn't look like the user preference we should be loading 🤔

The landing page preference on `/me` looks like it is handled by two separate calypso preferences: `reader-landing-page` and `sites-landing-page`

abdullah-kasim replied:
Thanks, I'll check the queries again. I crafted this PR without checking for previous behavior admittedly.

p-jackson commented on packages/api-core/src/me-preferences/types.ts:
```diff
@@ -9,7 +9,15 @@ export type SitesViewPreferences = Partial< Omit< SitesView, 'type' | 'layout' >
 };
 
+export type LoginPreferences = {
+	primarySiteId?: string;
```
The primary site is already stored as part of the user settings returned by `/me/settings` no?
https://github.com/Automattic/wp-calypso/blob/e355be00c8dc0e2a52e51ac68afc23464a1a9fec/packages/api-core/src/me-profile/fetchers.ts#L5

p-jackson commented on client/dashboard/me/preferences-login/index.tsx:
```diff
@@ -0,0 +1,191 @@
+	const { createSuccessNotice, createErrorNotice } = useDispatch( noticesStore );
+
+	// Fetch current login preferences
+	const { data: loginPrefs } = useSuspenseQuery( userPreferenceQuery( 'login-preferences' ) );
```
And data loaded using a `useSuspenseQuery` needs to be loaded in the router, otherwise the screen will just flash blank until the data is loaded.

For example, see the `purchasesRoute` loader. The purchase data is required in order to render the purchases view, and so it loads it first.
https://github.com/Automattic/wp-calypso/blob/e355be00c8dc0e2a52e51ac68afc23464a1a9fec/client/dashboard/app/router/me.tsx#L86

p-jackson commented on client/dashboard/me/preferences/test/index.test.tsx:
```diff
@@ -0,0 +1,110 @@
+	jest.resetAllMocks();
+} );
+
+test( 'renders preferences page with correct layout', async () => {
```
I'm not sure how much value these `<Preferences>` component tests add, since there isn't any sort of behaviour we're trying to verify 🤔. They're just a snapshot in time, is there much value in ensuring the view title doesn't change?

abdullah-kasim replied:
Hmm, I understand. There isn't much value to it - I was initially adding it as we want to check if the thing is rendering, but all the other tests already do that. Will make the change.

abdullah-kasim replied:
I contemplated a bit, checked the file name on where this test was coming from which is `me/preferences/test/index.test.tsx`.

I'm inclined to leave this in as it acts more like a snapshot test to see if somebody has accidentally removed this component from `preferences/index.tsx`, or accidentally made it not render. That's also why it's simply mocking `preferences-login` as what's inside doesn't really matter - I just want to make sure that we didn't forget to call it.

I'll change the test name to be more succinct with the intention, and elminate the unneeded ones. The above is really the only behavior I need tested.

p-jackson commented on client/dashboard/me/preferences-login/test/index.test.tsx:
```diff
@@ -0,0 +1,392 @@
+	expect( saveButton ).toBeEnabled();
+} );
+
+test( 'saves preferences successfully', async () => {
```
I like these `<PreferencesLogin>` tests though, where we're asserting some behaviour!

p-jackson reviewed (commented):
I think there's an issue here, because the settings in the v2 preferences should map to the settings which already exist on WordPress.com

For example, when I change the setting in `/me/account`, I expect the change to be reflected in `/v2/me/preferences`

https://github.com/user-attachments/assets/813bea87-17c0-42d1-971a-b94ab5aea28f

That might address the todo from the PR description: `Logic required to actually read the preferences and act on the preferences`

Because the logic for these settings is already in place.

andrea-sdl commented:
@abdullah-kasim besides what @p-jackson mentioned [here](https://github.com/Automattic/wp-calypso/pull/105389#pullrequestreview-3174869744) about the changes not being reflected (something I experienced too) I have some UI feedback too

I noticed in the original designs we have also an icon as well as the domain name
<img width="1224" height="112" alt="image" src="https://github.com/user-attachments/assets/5658c06e-bc05-40ea-aef3-2de0803b2f85" />

Plus, in the old logic, we would highlight the site type/visibility, as well as show that icon (which I think is the site icon) and, last but not least, a search option.
<img width="1242" height="1148" alt="CleanShot 2025-09-02 at 11 48 15@2x" src="https://github.com/user-attachments/assets/d5a0b28d-089c-4247-8d4f-173a09b7480b" />

Not sure if you're were planning to tackle this separately, but I thought it was good to mention since there is some UI differences.
I'm not even sure if we should consider a different component for this,.
Looking at the design system the `ComboboxControl` could be a good fit, as it has the search capability and [custom render for the list](https://wordpress.github.io/gutenberg/?path=/story/components-comboboxcontrol--with-custom-render-item&args=messages.selected:aaa), but it lacks a custom display for the selected value (something we do have in the WIP [CustomSelectControlV2](https://wordpress.github.io/gutenberg/?path=/story/components-customselectcontrol-v2--custom-selected-value) which, in turn, lacks a search 🤷 )

abdullah-kasim committed:
Merge remote-tracking branch 'origin/trunk' into add/preferences-login

abdullah-kasim committed:
Fix merging artifacts

abdullah-kasim committed:
Add the core api for the `me/settings` endpoint

abdullah-kasim committed:
Make the types smaller - we can keep adding more when necessary.

abdullah-kasim committed:
Add default values for landing page settings

abdullah-kasim committed:
Add more default values for me-preferences.ts

abdullah-kasim committed:
Add exports from api-queries

abdullah-kasim committed:
Add me-settings query

abdullah-kasim committed:
Update test to only do login page component detection

abdullah-kasim committed:
Refactored the login block to use pre-existing APIs. Let's gooo

abdullah-kasim committed:
Add login-specific queries

abdullah-kasim committed:
Add query.ts to minimize business logic inside view layer and switch
over to using the old way of doing things so that we can make use of
pre-existing redirect logic

abdullah-kasim committed:
Update tests

abdullah-kasim committed:
whoops, remove timeout added for debugging purposes

abdullah-kasim committed:
Use combobox for selecting a site

abdullah-kasim committed:
Use globe icon as the default icon to match previous calypso

abdullah-kasim committed:
Fix endless rerender on formData by having `setInit` set the value once,
and move view logic from query to the login block

abdullah-kasim committed:
Sites dropdown isn't complicated enough to deserve its own folder

abdullah-kasim committed:
Delete 'presence' tests as we focus on business logic

abdullah-kasim committed:
Fix warn deprecations

abdullah-kasim committed:
Remove more presence tests

abdullah-kasim committed:
Merge remote-tracking branch 'origin/trunk' into add/preferences-login

abdullah-kasim committed:
Remove duplicate route

andrea-sdl committed:
Merge branch 'trunk' into add/preferences-login

andrea-sdl committed:
fix readability of text in the dropdown

andrea-sdl committed:
make the language appear first, then the login options, as indicated by designs

andrea-sdl committed:
Switch to using SectionHeader for section titles

andrea-sdl committed:
invert "selected" icon color on hover

andrea-sdl committed:
remove extra menu item

andrea-sdl commented on client/dashboard/me/preferences-language/index.tsx:
```diff
@@ -215,7 +215,8 @@ export default function PreferencesLanguageForm() {
 				<CardBody>
-					<VStack spacing={ 6 } className="dasboard-preferences__vstack">
+					<VStack spacing={ 3 } className="dasboard-preferences__vstack">
+						<SectionHeader level={ 3 } title={ __( 'Language' ) } />
```
I noticed we're using `SectionHeader` in other parts, so I "upgraded" this page to use that component.

andrea-sdl committed:
fix type error in site-dropdown-test

fushar commented on client/dashboard/me/preferences-login/index.tsx:
```diff
@@ -0,0 +1,164 @@
+			setFormData( data );
+			setInit( true );
+		}
+	}, [ loginPreferences, init ] );
```
I think, instead of initializing the form in an `useEffect()` like this, we can simply preload the required queries, i.e. the preferences and user settings queries in the routes:

https://github.com/Automattic/wp-calypso/blob/220e6ffa78d0884d3fdbbd2a5539fd75bfd02716/client/dashboard/app/router/me.tsx#L80

(we just need to add `userSettingsQuery()`)

Then, we can use `useSuspenseQuery()` and with this we can make sure that data is already available when the component is rendered.

andrea-sdl replied:
@fushar There were multiple queries for that single page, I aggregated them all in the loader here https://github.com/Automattic/wp-calypso/pull/105389/commits/15bae21191a219bc5c8d84d97ff3b585d5c7078a what do you think?

fushar commented on client/dashboard/me/preferences-login/query.ts:
```diff
@@ -0,0 +1,113 @@
+		userPreferenceMutation( READER_AS_LANDING_PAGE_PREFERENCE )
+	);
+
+	return useMutation( {
```
I think it's better to create the mutation in `api-queries` instead, which then calls `api-core` endpoints. For example:

`api-queries/me-preferences-login.ts`:

```ts
export const userLoginPreferencesMutation( ...
```

So, we're composing multiple `api-core` calls inside a mutation, instead of composing mutations.

What do you think?

andrea-sdl replied:
@fushar  I refactored a little bit in https://github.com/Automattic/wp-calypso/pull/105389/commits/01ba37b089727b0d2b30ba431fab468e75ff173e so that most of the query/mutation is inside the newly created `me-preferences-login.ts` file. I still kept the `query.ts` at the `./client/dashboard` level because I liked how clean the UI code was, but it shouldn't be a big deal if we want to remove that file too and merge it in the main `index.tsx`.

andrea-sdl replied:
As you'll notice I also added a dedicated query to retrieve all the Preferences in one go and return the two ones we care about, instead of querying it twice.

fushar commented on client/dashboard/me/preferences-login/query.ts:
```diff
@@ -0,0 +1,113 @@
+			await readerLandingUpdate.mutateAsync( {
+				useReaderAsLandingPage: data.defaultLandingPage === 'reader',
+				updatedAt,
+			} );
```
It seems this `updatePreferences()` api-core function supports updating multiple keys 🤔 can we investigate whether we can use it and only call the endpoint once? 

https://github.com/Automattic/wp-calypso/blob/220e6ffa78d0884d3fdbbd2a5539fd75bfd02716/packages/api-core/src/me-preferences/mutators.ts#L4

andrea-sdl replied:
Circling back here: I didn't find any supporting evidence that we -need- to run the mutation twice. Looking at the [old code](https://github.com/Automattic/wp-calypso/blob/15bae21191a219bc5c8d84d97ff3b585d5c7078a/client/me/account/toggle-landing-page.tsx#L33-L44) it seems to me we had a function to retrieve a single preference value, and my guess is that we used that for each preference, even though we supported multiple ones.

I tested querying and saving them in the same API call and didn't notice any problem, for now we'll proceed with the consolidated approach in V2, using a single API call instead of two.

andrea-sdl replied:
ref: p1758193226775369-slack-C0982082MSR

andrea-sdl committed:
Merge branch 'trunk' into add/preferences-login

andrea-sdl committed:
preload queries before the userPreferences page to ensure the login settings form is ready without needing the useEffect

andrea-sdl committed:
create the me-preferences-login in api-queries to consolidate the query and the mutations required for the login-preferences

andrea-sdl committed:
Merge branch 'trunk' into add/preferences-login

andrea-sdl committed:
refactor: remove unused LandingPage type import in PreferencesLogin component

andrea-sdl committed:
Refactor API mocks in preferences-login tests for clarity

fushar reviewed (approved):
Tested the feature; I think it works well. Thanks for taking it to the finish line!

andrea-sdl committed:
Merge branch 'trunk' into add/preferences-login

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/20116120

Some locales (Hebrew, Japanese) have been temporarily machine-translated due to translator availability. All other translations are usually ready within a few days. Untranslated and machine-translated strings will be sent for translation next Monday and are expected to be completed by the following Friday.

Thank you @abdullah-kasim for including a screenshot in the description! This is really helpful for our translators.
